File: packages/code-storage-go/fetch.go
package storage

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/url"
	"strings"
)

type apiFetcher struct {
	baseURL    string
	version    int
	httpClient *http.Client
}

func newAPIFetcher(baseURL string, version int, client *http.Client) *apiFetcher {
	if client == nil {
		client = http.DefaultClient
	}
	return &apiFetcher{baseURL: strings.TrimRight(baseURL, "/"), version: version, httpClient: client}
}

func (f *apiFetcher) basePath() string {
	return f.baseURL + "/api/v" + itoa(f.version)
}

func (f *apiFetcher) buildURL(path string, params url.Values) string {
	if params == nil || len(params) == 0 {
		return f.basePath() + "/" + path
	}
	return f.basePath() + "/" + path + "?" + params.Encode()
}

type requestOptions struct {
	allowedStatus map[int]bool
	// extraHeaders merges request headers on top of SDK defaults; empty values are skipped.
	extraHeaders map[string]string
}

func (f *apiFetcher) request(ctx context.Context, method string, path string, params url.Values, body interface{}, jwt string, opts *requestOptions) (*http.Response, error) {
	if ctx == nil {
		ctx = context.Background()
	}

	urlStr := f.buildURL(path, params)
	var bodyReader io.Reader
	if body != nil {
		payload, err := json.Marshal(body)
		if err != nil {
			return nil, err
		}
		bodyReader = bytes.NewReader(payload)
	}

	req, err := http.NewRequestWithContext(ctx, method, urlStr, bodyReader)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Authorization", "Bearer "+jwt)
	req.Header.Set("Code-Storage-Agent", userAgent())
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	if opts != nil {
		for k, v := range opts.extraHeaders {
			if v != "" {
				req.Header.Set(k, v)
			}
		}
	}

	resp, err := f.httpClient.Do(req)
	if err != nil {
		return nil, err
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		if opts != nil && opts.allowedStatus != nil && opts.allowedStatus[resp.StatusCode] {
			return resp, nil
		}

		defer resp.Body.Close()
		bodyBytes, _ := io.ReadAll(resp.Body)
		var parsed interface{}
		message := ""
		contentType := resp.Header.Get("content-type")
		if strings.Contains(contentType, "application/json") {
			var payload map[string]interface{}
			if err := json.Unmarshal(bodyBytes, &payload); err == nil {
				parsed = payload
				if errVal, ok := payload["error"].(string); ok && strings.TrimSpace(errVal) != "" {
					message = strings.TrimSpace(errVal)
				}
			}
		}
		if message == "" && len(bodyBytes) > 0 {
			message = strings.TrimSpace(string(bodyBytes))
			if message != "" {
				parsed = message
			}
		}

		if message == "" {
			message = "request " + method + " " + urlStr + " failed with status " + itoa(resp.StatusCode) + " " + resp.Status
		}

		return nil, &APIError{
			Message:    message,
			Status:     resp.StatusCode,
			StatusText: resp.Status,
			Method:     method,
			URL:        urlStr,
			Body:       parsed,
		}
	}

	return resp, nil
}

func (f *apiFetcher) get(ctx context.Context, path string, params url.Values, jwt string, opts *requestOptions) (*http.Response, error) {
	return f.request(ctx, http.MethodGet, path, params, nil, jwt, opts)
}

func (f *apiFetcher) head(ctx context.Context, path string, params url.Values, jwt string, opts *requestOptions) (*http.Response, error) {
	return f.request(ctx, http.MethodHead, path, params, nil, jwt, opts)
}

func (f *apiFetcher) post(ctx context.Context, path string, params url.Values, body interface{}, jwt string, opts *requestOptions) (*http.Response, error) {
	return f.request(ctx, http.MethodPost, path, params, body, jwt, opts)
}

func (f *apiFetcher) put(ctx context.Context, path string, params url.Values, body interface{}, jwt string, opts *requestOptions) (*http.Response, error) {
	return f.request(ctx, http.MethodPut, path, params, body, jwt, opts)
}

func (f *apiFetcher) delete(ctx context.Context, path string, params url.Values, body interface{}, jwt string, opts *requestOptions) (*http.Response, error) {
	return f.request(ctx, http.MethodDelete, path, params, body, jwt, opts)
}

File: packages/code-storage-typescript/src/fetch.ts
import { errorEnvelopeSchema } from './schemas';
import type { ValidAPIVersion, ValidMethod, ValidPath } from './types';
import { getUserAgent } from './version';

interface RequestOptions {
  allowedStatus?: number[];
  /** Extra request headers merged on top of the SDK defaults. */
  extraHeaders?: Record<string, string>;
  /** Resolve paths from `/api` instead of `/api/v{version}`. */
  apiRoot?: boolean;
  signal?: AbortSignal;
}

export class ApiError extends Error {
  public readonly status: number;
  public readonly statusText: string;
  public readonly method: ValidMethod;
  public readonly url: string;
  public readonly body?: unknown;

  constructor(params: {
    message: string;
    status: number;
    statusText: string;
    method: ValidMethod;
    url: string;
    body?: unknown;
  }) {
    super(params.message);
    this.name = 'ApiError';
    this.status = params.status;
    this.statusText = params.statusText;
    this.method = params.method;
    this.url = params.url;
    this.body = params.body;
  }
}

export class ApiFetcher {
  constructor(
    private readonly API_BASE_URL: string,
    private readonly version: ValidAPIVersion
  ) {}

  private getBaseUrl() {
    return `${this.API_BASE_URL}/api/v${this.version}`;
  }

  private getRequestUrl(path: ValidPath, apiRoot: boolean) {
    const baseUrl = apiRoot ? `${this.API_BASE_URL}/api` : this.getBaseUrl();
    if (typeof path === 'string') {
      return `${baseUrl}/${path}`;
    } else if (path.params) {
      const searchParams = new URLSearchParams();
      for (const [key, value] of Object.entries(path.params)) {
        if (Array.isArray(value)) {
          for (const v of value) {
            searchParams.append(key, v);
          }
        } else {
          searchParams.append(key, value);
        }
      }
      const paramStr = searchParams.toString();
      return `${baseUrl}/${path.path}${paramStr ? `?${paramStr}` : ''}`;
    } else {
      return `${baseUrl}/${path.path}`;
    }
  }

  private async fetch(
    path: ValidPath,
    method: ValidMethod,
    jwt: string,
    options?: RequestOptions
  ) {
    const requestUrl = this.getRequestUrl(path, options?.apiRoot === true);

    const headers: Record<string, string> = {
      Authorization: `Bearer ${jwt}`,
      'Content-Type': 'application/json',
      'Code-Storage-Agent': getUserAgent(),
    };
    if (options?.extraHeaders) {
      for (const [key, value] of Object.entries(options.extraHeaders)) {
        if (typeof value === 'string' && value !== '') {
          headers[key] = value;
        }
      }
    }

    const requestOptions: RequestInit = {
      method,
      headers,
      signal: options?.signal,
    };

    if (
      method !== 'GET' &&
      method !== 'HEAD' &&
      typeof path !== 'string' &&
      path.body
    ) {
      requestOptions.body = JSON.stringify(path.body);
    }

    const response = await fetch(requestUrl, requestOptions);

    if (!response.ok) {
      const allowed = options?.allowedStatus ?? [];
      if (allowed.includes(response.status)) {
        return response;
      }

      let errorBody: unknown;
      let message: string | undefined;
      const contentType = response.headers.get('content-type') ?? '';

      try {
        if (contentType.includes('application/json')) {
          errorBody = await response.json();
        } else {
          const text = await response.text();
          errorBody = text;
        }
      } catch {
        // Fallback to plain text if JSON parse failed after reading body
        try {
          errorBody = await response.text();
        } catch {
          errorBody = undefined;
        }
      }

      if (typeof errorBody === 'string') {
        const trimmed = errorBody.trim();
        if (trimmed) {
          message = trimmed;
        }
      } else if (errorBody && typeof errorBody === 'object') {
        const parsedError = errorEnvelopeSchema.safeParse(errorBody);
        if (parsedError.success) {
          const trimmed = parsedError.data.error.trim();
          if (trimmed) {
            message = trimmed;
          }
        }
      }

      throw new ApiError({
        message:
          message ??
          `Request ${method} ${requestUrl} failed with status ${response.status} ${response.statusText}`,
        status: response.status,
        statusText: response.statusText,
        method,
        url: requestUrl,
        body: errorBody,
      });
    }
    return response;
  }

  async get(path: ValidPath, jwt: string, options?: RequestOptions) {
    return this.fetch(path, 'GET', jwt, options);
  }

  async head(path: ValidPath, jwt: string, options?: RequestOptions) {
    return this.fetch(path, 'HEAD', jwt, options);
  }

  async post(path: ValidPath, jwt: string, options?: RequestOptions) {
    return this.fetch(path, 'POST', jwt, options);
  }

  async put(path: ValidPath, jwt: string, options?: RequestOptions) {
    return this.fetch(path, 'PUT', jwt, options);
  }

  async patch(path: ValidPath, jwt: string, options?: RequestOptions) {
    return this.fetch(path, 'PATCH', jwt, options);
  }

  async delete(path: ValidPath, jwt: string, options?: RequestOptions) {
    return this.fetch(path, 'DELETE', jwt, options);
  }
}
